lib/chef/provisioning/fog_driver/providers/softlayer.rb

```diff
@@ -28,6 +34,96 @@ def self.compute_options_for(provider, id, config)
 
             [result, id]
           end
+
+          def bootstrap_options_for(action_handler, machine_spec, machine_options)
+            opts = super
+            if opts[:bare_metal]
+              # these don't work with fog-softlayer bare_metal for some reason...
+              # NOTE: bare metal support here is not fully tested and probably
+              # won't work anyway
+              opts.delete :key_name
+              opts.delete :tags
+            end
+
+            # we hook in our own post-install script which uses userMetadata to
+            # tell us when post-install is complete. If the user supplies their
+            # own script it will be called by our hook before indicating
+            # completion in userData.
+            opts[:postInstallScriptUri] = 'https://dal05.objectstorage.service.networklayer.com/v1/AUTH_b1b23a05-1c03-4961-8b08-2339886e476f/dist/sl-post-hook.sh'
+
+            opts
+          end
+
+          def find_floating_ips(server, action_handler)
+            []
+          end
+
+          def create_servers(action_handler, specs_and_options, parallelizer, &block)
+            super do |machine_spec, server|
+              bootstrap_options = specs_and_options[machine_spec][:bootstrap_options]
+              set_post_install_info(server, bootstrap_options)
+
+              block.call(machine_spec, server) if block
+            end
+          end
+
+          def set_post_install_info(server, bootstrap_options)
+            existing_user_data = server.service.request( :virtual_guest, server.id, :query => {:objectMask => 'userData'}).body['userData']
+            existing_user_data ||= ''
+            if existing_user_data.empty?
+              creds = Fog.credentials
+              ::Retryable.retryable(:tries => 60, :sleep => 2) do
+                sleep(2)
+
+                update_url = URI::HTTPS.build(
+                  :userinfo => "#{creds[:softlayer_username]}:#{creds[:softlayer_api_key]}",
+                  :host => 'api.service.softlayer.com',
+                  :path => "/rest/v3/SoftLayer_Virtual_Guest/#{server.id}/setUserMetadata",
```

Comment:
If you were going to work on supporting bare metal servers you'd need to make this conditional.

```ruby
"/rest/v3/#{opts.bare_metal ? SoftLayer_Hardware_Server : SoftLayer_Virtual_Guest}/#{server.id}/setUserMetadata"
```

or perhaps something equivalent that's a tad more readable...